bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/Display.java

```diff
@@ -5152,16 +5152,90 @@ String wrapText (String text, long handle, int width) {
 }
 
 static String withCrLf (String string) {
-	if (string == null) {
+
+	/* If the string is empty, return the string. */
+	int length = string.length ();
+	if (length == 0) return string;
+
+	/*
+	* Check for an LF or CR/LF and assume the rest of
+	* the string is formated that way.  This will not
+	* work if the string contains mixed delimiters.
+	*/
+	int i = string.indexOf ('\n', 0);
+	if (i == -1) return string;
+	if (i > 0 && string.charAt (i - 1) == '\r') {
 		return string;
 	}
-	// Replace \r\n, \r, or \n with \r\n
-	return string.replaceAll("(\r\n|\r|\n)", "\r\n");
+
+	/*
+	* The string is formatted with LF.  Compute the
+	* number of lines and the size of the buffer
+	* needed to hold the result
+	*/
+	i++;
+	int count = 1;
+	while (i < length) {
+		if ((i = string.indexOf ('\n', i)) == -1) break;
+		count++; i++;
+	}
+	count += length;
+
+	/* Create a new string with the CR/LF line terminator. */
+	i = 0;
+	StringBuilder result = new StringBuilder (count);
+	while (i < length) {
+		int j = string.indexOf ('\n', i);
+		if (j > 0 && string.charAt(j - 1) == '\r') {
+			result.append(string.substring(i, j + 1));
+			i = j + 1;
+		} else {
+			if (j == -1) j = length;
+			result.append (string.substring (i, j));
+			if ((i = j) < length) {
+				result.append ("\r\n"); //$NON-NLS-1$
+				i++;
+			}
+		}
+	}
+	return result.toString ();
```

Comment:
should be save to avoid a copy with `if (string.length== result.length()) return string;`

Reply:
Done